### ISSUE TITLE
OOM Investigation 

### ISSUE THREAD

**Issue by rosibaj** [icgc-argo/workflow-roadmap#136]
- @lindaxiang please provide a list of OOM Sanger jobs that have failed to investigate 
- CPU/mem logs are available from Nextflow.  we need to verify if we are retrieving those, or alternately check the pod logs for information. 
- establish a pattern of behavior in these new Sanger WGS runs that are causing a high rate of OOM failures
  - cpui parameters were lowered on resume; verify if this solved the issue for resume 
  - if needed, rerun previously successful jobs to establish if the issue occurs on previously successful data 
  - triangulation between WF upgrade, Kube Upgrade could be factors here

**Comment by lindaxiang** [comment]
Here is the list of sanger jobs which failed because of out of memory issue.
```
wes-4a9491310ad740ed9321e79090a1e2bd
wes-f0183b59b3504479ab402ab1e560d0ea
wes-236a49526ccf4d8190a0d3373b72ca9d
wes-d9760ab286744b03ac0b9c56d5311ba3
wes-6d85bb355f14427d981c2aaec7f52273
```
FYI. the OOM issue often get solved if lower the cpu parameters.

**Comment by lepsalex** [comment]
The Mystery of the OOM pods, by Alex Lepsa ... the first thing to know when diving into this mystery is how our main protagonist, Nextflow, schedules pods; they always set a resource request to be equal to a resource limit. This means that requesting 170gb of memory, as is the case when running the main Sanger process, means that the pod requests 170gn and is limited to 170gb.

Now the Sanger process doesn't allow you to specify a memory max (to my knowledge), instead they ask you how many CPUs you'd like to use and then those CPUs each slurp up memory as needed, and while Sanger claims that the max is 6GB per CPU we have the resource usage graphs to prove that it often goes over 7gb. Usually this isn't a problem since we have requested (and limited) at a ratio that mostly gives each core the memory it needs. That is except for these OOM pods, where they appear to be running out anyways, and so a mystery is born.

Well dear reader I hate to disappoint you but this mystery is a rather simple one to solve. After looking at some overall memory usage on the nodes where these pods ran showed that they still had plenty of memory available, we consulted with the great oracle of Kubernetes, the [docs site](https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/#exceed-a-container-s-memory-limit), which I shall quote for you now, "A Container can exceed its memory request if the Node has memory available. But a Container is not allowed to use more than its memory limit. **If a Container allocates more memory than its limit, the Container becomes a candidate for termination.**"

Due to the fact that the Sanger job allows each core to use as much memory as needed for their process, in the event of what I suppose is more taxing variant calling, the pod exceeds it's limits and Kubernetes comes down with the wrath of the heavens resulting in a OOMKilled pod status.

There is nothing we can do about this on the infra side, the 170gb memory limit was chosen to maximize the amount of eligible nodes for running Sanger jobs, so instead what I am suggesting is that we use the built-in features of Nextflow to retry these more troublesome of variant callers with progressively lower CPU requests, which in turn should use less memory and should eventually succeed without our manual intervention.

Details about this can be found here: [Nextflow Dynamic Resource](https://www.nextflow.io/docs/latest/process.html#dynamic-computing-resources), I would suggest something like this:

```
process sangerWgsVariantCaller {
    cpus {task.attempt > 1 ? params.cpus - (parms.cpu_backoff * task.attempt) : params.cpus}

    errorStrategy { task.exitStatus in 137..140 ? 'retry' : 'terminate' }
    maxRetries params.max_retries
}
```

Basically on attempt one we use the full requested CPU amount from the params, then progressively lower by the `cpu_backoff` param (**this would be new**) until we've retried `max_retries` times (**also new**). 

@lindaxiang @rosibaj @christinayung

**Comment by keiranmraine** [comment]
Hi, one of the places where memory an occasionally get a bit out of hand is while cgpPindel and caveman are running some elements in parallel (parallel block 3).

You can often get a reliable reduction in memory by dropping the pindel cores to 4, it may be worth doing this for for all your jobs:

https://github.com/cancerit/dockstore-cgpwgs/blob/50fabe3c9b72d885d4010678ee1dc92c379fbf84/cwls/cgpwgs.cwl#L137-L143

If you are able to inspect the work directory and identify where in the pindel process it is at when it fails you can confirm if this will help:

```bash
$ ls -ltrh $(find $ROOT_OF_OUTPUT -type d -path */tmpPindel/progress)
```

Internally we use containers for the individual algorithms so we don't generally hit this issue.

**Comment by keiranmraine** [comment]
... internally we run caveman mstep/estep with 8GB/core since the migration towards GRCh38.

**Comment by lindaxiang** [comment]
Currently in production the default setting for sanger wgs calling:
```
  "mem": 170,
  "cpus": 24,
  "pindelcpu": 3
```

**Comment by keiranmraine** [comment]
As you only have 170GB ram available when using hosts with 24 CPU you may be better setting the default at 20 cpus.  For a significant amount of the time caveman is running pindel is also running, the 3 CPUs for pindel are not deducted from the total.

**Comment by lindaxiang** [comment]
Thanks @keiranmraine for the information. As the peak memory usage for two parallel running tasks could still go high as (20+3)*8=184 > 170GB, I was thinking to decrease the `cpus=18`.  Is my understanding correct?

**Comment by keiranmraine** [comment]
Your understanding is correct, but the pindel memory usage isn't the same as the caveman.  It's a bit of a trade off on how much CPU you are happy to have idle vs repeated jobs with a back-off.

For back-off I'd always drop 4 cpus at a time.
